site/content/contribute/more-info/server/schema-migration-guide.md

```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
 
-The problem arises when in some databases, for some tables, due to various technical reasons, applying a migration prevents other operations from happening on that table. In that case, it causes unavoidable downtime.
+Adding an FK takes locks. But avoidable by setting `SET  foreign_key_checks=0`. Then it doesn't take any locks, and neither rebuilds the table. But it partially defeats the purpose of using an FK. Recommend to AVOID.
 
-### TL;DR
+6. DROP INDEX
 
-1. Migrations are always synchronous.
-2. The time taken to run a migration is not always the primary factor in judging impact.
-3. Except in cases where a migration cannot run concurrently with other database operations.
+**Postgres**: DROP INDEX CONCURRENTLY does not take any locks.
+
+**MySQL**: Fast operation. No locks are taken.
+
+7. CREATE TABLE
+
+Does not lock any existing data so no issues.
+
+8. UPDATE
+
+An analysis shows that UPDATE statements roughly fall into one of these three categories:
+
+- **Data migrations**:
+
+```sql
+UPDATE channelmembers SET MentionCountRoot = ..
+UPDATE Channels SET TotalMsgCountRoot = ..
+UPDATE ChannelMembers CM SET MsgCountRoot ..
+```
+
+In these cases, rather than operating on the entire table, we need to operate on batches at a time. See the example later on how to achieve that.
+
+- **Changing NULL columns to NON-NULL**
+
+```sql
+UPDATE Channels SET LastRootPostAt=0 WHERE LastRootPostAt IS NULL;
+UPDATE OAuthApps SET MattermostAppID = '' WHERE MattermostAppID IS NULL;
+```
+
+This is possible to handle from the code itself using a `COALESCE` function. It makes the code more complicated, but it’s a cost we have to pay to reduce migration overhead.
+
+- **Denormalization optimizations**:
+
+```sql
+UPDATE threads SET threaddeleteat = posts.deleteat FROM posts WHERE threads.threaddeleteat IS NULL AND posts.id = threads.postid;
+UPDATE reactions SET channelid = COALESCE((select channelid from posts where posts.id = reactions.postid), '') WHERE channelid='';
+UPDATE threads SET threadteamid = channels.teamid FROM channels WHERE threads.threadteamid IS NULL AND channels.id = threads.channelid;
+UPDATE fileinfo SET channelid = posts.channelid FROM posts WHERE fileinfo.channelid IS NULL AND fileinfo.postid = posts.id;
+```
+
+We can take the same approach as in data migrations.
+
+9. DELETE
+
+So far, there have been only a handful of DELETE statements in schema migrations. And mostly they have been for security issues. The general recommendation is to avoid running a full-blown DELETE statement that operates on the entire table, but rather operate on batches so as to avoid taking a lock on the entire table. This could either be done in a job since there is no new code waiting for this to be executed. (See above)
+
+## Examples
+
+### How do I change a column type if I MUST
+
+Follow this long-winded procedure:
+- Create a new column.
+- Migrate existing data.
+- From next ESR, start using the new column.
+- Next ESR, drop the old column.
```

Comment:
What do you think about helping readers by adding a visual for this? Maybe it can be easier to understand.

Reply:
Sure, let me add a diagram